### DISCUSSION TITLE
Company stops working in Eshell after some commands

### DISCUSSION THREAD

**Discussion by MintSoup** [company-mode/company-mode#1276] [Q&A]
I have company-mode enabled in Eshell, but after running some commands, company stops displaying the selection dialog.
For example, if I type "pacman" in eshell, and press enter (without selecting a suggestion), company will stop suggesting options afterwards. However, if I type "pacman" again but this time select one of the suggested options, company will keep working fine. This happens with other programs too but this is the only case I have been able to consistently reproduce. Also note that whenever company stops working in Eshell, it doesn't come back even if I kill the buffer and open another eshell, I have to restart Emacs for it to come back. It works fine in other buffers though. Any ideas?

**Comment by dgutov** [comment]
I'm guessing in certain cases, for certain inputs, its completion function misbehaves and gets put in `completion--capf-misbehave-funs` by `completion--capf-wrapper`.

Do you ever see the message "Completion function %S uses a deprecated calling convention" in Messages? You can double-check with `C-h e`.

Also see the value of `completion-at-point-functions` inside an Eshell buffer, to find out which completion function we are talking about.

**Reply by MintSoup** [reply]
Thank you for the quick reply.

Yes, I get `Completion function pcomplete-completions-at-point uses a deprecated calling convention`

```
~ $ echo $completion-at-point-functions
(pcomplete-completions-at-point t)
```

**Reply by dgutov** [reply]
But not every time, right? Not right away when you use completion in Eshell?

If it happens only in certain context (with certain text on the current line, for instance), that can be a reproduction scenario for a bug report.

One you (or me) could eventually file against Emacs, that is. If it's still present in the latest development snapshot.

**Reply by MintSoup** [reply]
It doesn't happen with every command but for commands that do cause it it happens every time. I did some digging and found these threads:
https://github.com/company-mode/company-mode/issues/409
https://github.com/hlissner/doom-emacs/issues/3817

From what I understand this is an old bug, but I couldn't get Doom's workaround to work for me.
What do people usually use to get "normal" eshell completions with company?

**Reply by dgutov** [reply]
Ah, indeed. Typing a quote does that to it.

I've added this bit of information to https://debbugs.gnu.org/cgi/bugreport.cgi?bug=50470, which is a previous report on this issue.

As far as recommendations go, there are a couple in the messages on the linked page, but the result will be that you get much worse completion in eshell (or basically none). You could try disabling `company-capf` in those buffers, though. Then you'd get advantage of `company-dabbrev` or `company-dabbrev-code`.

**Reply by MintSoup** [reply]
I managed to get completions working decently using [this.](https://xenodium.com/eshell-pcomplete-company-completion/)

**Comment by dgutov** [comment]
This should be fixed now on master.

**Reply by MintSoup** [reply]
Yup, seems to work. Thanks!
